### PR TITLE
Extract program-error crate

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Problem
`solana_program::program_error` needs to be made available outside solana-program

This branches off #2405 so that needs to be merged first.

#### Summary of Changes
- move program_error.rs to its own crate and re-export its contents for backwards compatibility
- make serde optional in the new crate
- remove thiserror from the new crate

### CODE REVIEW

File: sdk/program/src/entrypoint.rs [hunk 1]
```diff
@@ -5,22 +5,20 @@
 //! [`bpf_loader`]: crate::bpf_loader
 
 extern crate alloc;
+pub use solana_program_error::ProgramResult;
```
Comment (joncinque) [inline sdk/program/src/entrypoint.rs:8]: Gosh, this makes so much more sense

File: sdk/program-error/src/lib.rs
Comment (joncinque) [inline sdk/program-error/src/lib.rs:1]: I agree with your approach of *not* making this no-std -- if we ever want to do it, we can introduce a breaking change later.

File: sdk/program-error/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+             => f.write_str("An attempt to operate on an account that hasn't been initialized"),
+            ProgramError::NotEnoughAccountKeys
+             => f.write_str("The instruction expected additional account keys"),
+            ProgramError::AccountBorrowFailed
+             => f.write_str("Failed to borrow a reference to account data, already borrowed"),
+            ProgramError::MaxSeedLengthExceeded
+             => f.write_str("Length of the seed is too long for address generation"),
+            ProgramError::InvalidSeeds
+             => f.write_str("Provided seeds do not result in a valid address"),
+            ProgramError::BorshIoError(s) =>  write!(f,"IO Error: {s}"),
```
Comment (joncinque) [inline sdk/program-error/src/lib.rs]: micro-nit: I'm surprised rustfmt didn't do anything about this
```suggestion
            ProgramError::BorshIoError(s) =>  write!(f, "IO Error: {s}"),
```

**Review by joncinque — COMMENTED** [review]
Looks good to me! Just the one missing space, but it looks like rustfmt doesn't care. @yihau can you accept `solana-program-error`?

**Comment by yihau** [comment]
`crate-check` passed!

**Review by joncinque — APPROVED** [review]

**Pull request merged**